### PR TITLE
fix: Crash in SentryOutOfMemoryScopeObserver

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Writing data to disk when the file handle is broken causes crashes. This is fixed now by capturing an exception.

## :bulb: Motivation and Context

Fixes GH-2548

## :green_heart: How did you test it?
Unit tests.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

File: Samples/TrendingMovies/TrendingMovies/ImageProcessing/ColorArt.swift [hunk 1]
```diff
@@ -319,7 +319,7 @@ private struct RGBADecimalComponents: Hashable {
 
     // Relative luminance formula: https://en.wikipedia.org/wiki/Relative_luminance
     var luminance: CGFloat {
-        0.212_6 * r + 0.715_2 * g + 0.072_2 * b
+        0.2126 * r + 0.7152 * g + 0.0722 * b
```
Comment (philipphofmann) [inline Samples/TrendingMovies/TrendingMovies/ImageProcessing/ColorArt.swift:322]: Sorry about these unrelated code formatting changes. SwiftLint changed and we didn't run it on master with the latest version yet. I'm being pragmatic here to not open up a tiny PR this time.

Comment (armcknight) [reply Samples/TrendingMovies/TrendingMovies/ImageProcessing/ColorArt.swift:322]: i thought these were already in the 8.0.0 branch?   https://github.com/getsentry/sentry-cocoa/pull/2504#issuecomment-1344011248

Comment (philipphofmann) [reply Samples/TrendingMovies/TrendingMovies/ImageProcessing/ColorArt.swift:322]: This is a hotfix for the master branch.

File: Sources/Sentry/SentryOutOfMemoryScopeObserver.m [hunk 2]
```diff
@@ -75,15 +75,19 @@ - (void)switchFileHandle
...
+    @try {
+        [self.fileHandle seekToEndOfFile];
+        [self.fileHandle writeData:data];
+        [self.fileHandle writeData:[@"\n" dataUsingEncoding:NSASCIIStringEncoding]];
+
+        self.breadcrumbCounter += 1;
+
+        if (self.breadcrumbCounter >= self.maxBreadcrumbs) {
+            [self switchFileHandle];
+            self.breadcrumbCounter = 0;
```
Comment (armcknight) [inline Sources/Sentry/SentryOutOfMemoryScopeObserver.m:87]: This seems like it should go into a `@finally` block, unless we don’t want to reset the count of an exception occurs? Even switching file handles might help the next try work.

File: Sources/Sentry/SentryOutOfMemoryScopeObserver.m [hunk 3]
```diff
@@ -75,15 +75,19 @@ - (void)switchFileHandle
...
-    [self.fileHandle writeData:[@"\n" dataUsingEncoding:NSASCIIStringEncoding]];
-
-    self.breadcrumbCounter += 1;
-
-    if (self.breadcrumbCounter >= self.maxBreadcrumbs) {
-        [self switchFileHandle];
-        self.breadcrumbCounter = 0;
+    @try {
+        [self.fileHandle seekToEndOfFile];
+        [self.fileHandle writeData:data];
```
Comment (armcknight) [inline Sources/Sentry/SentryOutOfMemoryScopeObserver.m:80]: We should use the newer one with an inout error in addition to wrapping in try, since this one’s deprecated and the error parameter might catch additional info or be used instead of an exception, which would help performance.

File: Sources/Sentry/SentryOutOfMemoryScopeObserver.m [hunk 4]
```diff
@@ -75,15 +75,19 @@ - (void)switchFileHandle
...
-    [self.fileHandle writeData:data];
-    [self.fileHandle writeData:[@"\n" dataUsingEncoding:NSASCIIStringEncoding]];
-
-    self.breadcrumbCounter += 1;
-
-    if (self.breadcrumbCounter >= self.maxBreadcrumbs) {
-        [self switchFileHandle];
-        self.breadcrumbCounter = 0;
+    @try {
+        [self.fileHandle seekToEndOfFile];
```
Comment (armcknight) [inline Sources/Sentry/SentryOutOfMemoryScopeObserver.m:79]: Let’s capture the return value and put it in the error log if it fails, the extra contract might help later.

**Review by armcknight — COMMENTED** [review]
Mostly looks fine, but I do wonder about how the breadcrumb counter is maintained. The other comments are not blockers, just suggestions.

File: Sources/Sentry/SentryOutOfMemoryScopeObserver.m (hunk 3 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryOutOfMemoryScopeObserver.m:80]: Yes, but this one only works on iOS 13.0 and above, see https://developer.apple.com/documentation/foundation/nsfilehandle/3172533-writedata

Comment (philipphofmann) [reply Sources/Sentry/SentryOutOfMemoryScopeObserver.m:80]: I would create a follow-up issue for using the new APIs, is that fine for you?

File: Sources/Sentry/SentryOutOfMemoryScopeObserver.m (hunk 2 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryOutOfMemoryScopeObserver.m:87]: Yep good catch 💯

File: Sources/Sentry/SentryOutOfMemoryScopeObserver.m (hunk 3 above)
Comment (armcknight) [reply Sources/Sentry/SentryOutOfMemoryScopeObserver.m:82]: Sure, that’s fine!

**Review by armcknight — APPROVED** [review]
